### DISCUSSION TITLE
Deterministic Data Pre-processing

### DISCUSSION THREAD

**Discussion by moonforsun** [Project-MONAI/MONAI#5977] [Q&A] [answered]
I am wondering how to make the MONAI's random transforms deterministic? 
I am using RandCropByLabelClassesd in my transforms. But everytime the random cropped data is different with the same global seed. The `seed_everything` is called from `pytorch_lightning`

Besdies the following setting, what else should I add to ensure the deterministic training?

    seed_everything(cfg.seed, workers=True)
    torch.backends.cudnn.determinstic = True
    torch.backends.cudnn.benchmark = False

Thank you in advance!

**Comment by KumoLiu** [comment, marked as answer]
Hi @moonforsunm,
You could use `set_determinism` in MONAI. You could also refer to any [tutorials](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb), most of them have this part.
```
from monai.utils import set_determinism
set_determinism(seed=0)
```

Hope it can help you, thanks!

**Reply by moonforsun** [reply]
This line indeed ensures the deterministic training although the documentation itself indicates this function will not affect the randomizable objects in `monai.transforms.Randomizable` (https://docs.monai.io/en/stable/transforms.html#monai.transforms.Randomizable).

Thank you very much!

**Reply by KumoLiu** [reply]
Hi @moonforsun, I think the note in `set_determinism` means that if you are using [`monai.data.DataLoader`](https://docs.monai.io/en/stable/data.html#monai.data.DataLoader), it will automatically set the seeds according to the global random state. Therefore just as Wenqi mentioned, if you are not using `monai.data.DataLoader`, you need to set the transform' random state by yourself.
https://docs.monai.io/en/stable/utils.html#monai.utils.misc.set_determinism

Hope I have explained it clearly. Thanks!

**Comment by wyli** [comment]
if you are not using `monai.data.DataLoader`, you should set the transform's random state as well:
```
crop = RandCropByLabelClassesd(keys=..., ...)
crop.set_random_state(cfg.seed)
```

**Reply by moonforsun** [reply]
Thank you for your reply. Yes, I did indeed use MONAI's Dataloader rather than the torch's Dataloader when I reported the issue.
